apps/src/javalab/Neighborhood.js

```diff
@@ -41,68 +37,80 @@ export default class Neighborhood {
 
     const slider = document.getElementById('slider');
     this.speedSlider = new Slider(10, 35, 130, slider);
-    this.lastSignalEndTime = null;
+    this.signals = [];
+    this.nextSignalIndex = 0;
   }
 
   handleSignal(signal) {
-    // Calculate time to wait until executing this signal. Animated steps such as move
-    // have an additional timeForSignal that is used by maze to animate the sprite. All
-    // signals have a pause between signals. The timing is multiplied by the pegmanSpeedMultiplier
-    // to speed up or slow down the overall animation.
-    let timeout = 0;
-    const timeForSignal =
-      this.getAnimationTime(signal) * this.getPegmanSpeedMultiplier();
-    const totalSignalTime =
-      timeForSignal + PAUSE_BETWEEN_SIGNALS * this.getPegmanSpeedMultiplier();
-    if (this.lastSignalEndTime) {
-      timeout = this.lastSignalEndTime - Date.now();
-      if (timeout < 0) {
-        timeout = 0;
+    // add next signal to our queue of signals
+    this.signals.push(signal);
+  }
+
+  // Process avaiable signals recursively. We process recursively to ensure
+  // the commands appear sequential to the user and all commands stay in sync.
+  processSignals() {
+    // if there is at least one signal we have not processed, process it
+    if (this.signals.length > this.nextSignalIndex) {
+      const signal = this.signals[this.nextSignalIndex];
+      if (signal.value === NeighborhoodSignalType.DONE) {
+        // we are done processing commands and can stop checking for signals
+        return;
       }
-      this.lastSignalEndTime = this.lastSignalEndTime + totalSignalTime;
+      const timeForSignal =
+        this.getAnimationTime(signal) * this.getPegmanSpeedMultiplier();
+      const totalSignalTime =
+        timeForSignal + PAUSE_BETWEEN_SIGNALS * this.getPegmanSpeedMultiplier();
+      this.mazeCommand(signal, timeForSignal);
+
+      this.nextSignalIndex++;
+      // check for another signal after this signal has completed processing
+      timeoutList.setTimeout(() => this.processSignals(), totalSignalTime);
```

Comment:
Would it make sense to skip the wait if we're already lagging? i.e. Maybe something like this:
```
const totalSignalTime = timeForSignal + PAUSE_BETWEEN_SIGNALS * this.getPegmanSpeedMultiplier();
const beginTime = clock.now();
this.mazeCommand(signal, timeForSignal);
this.nextSignalIndex++;
const remainingTime = totalSignalTime - (clock.now() - beginTime)
timeoutList.setTimeout(() => this.processSignals(), MAX(totalSignalTime, 0));
```
I don't think it would really affect the faster execution speeds, but if it's running at turtle speed, it might be nice to avoid some extra waiting.

Reply:
this is a good call--it's most helpful for the big painting slowdown, because we aren't amplifying the slowdown by adding additional wait between actions. It shaves about 15 seconds off the animation (out of 1:26).